core/matrix/dense.cpp

```diff
@@ -637,123 +674,188 @@ void Dense<ValueType>::write(mat_data32 &data) const
-    auto trans_cpy = Dense::create(exec, gko::transpose(this->get_size()));
+    auto result =
+        Dense::create(this->get_executor(), gko::transpose(this->get_size()));
+    this->transpose(result.get());
+    return result;
+}
 
-    exec->run(dense::make_transpose(this, trans_cpy.get()));
 
-    return std::move(trans_cpy);
+template <typename ValueType>
+std::unique_ptr<LinOp> Dense<ValueType>::conj_transpose() const
+{
+    auto result =
+        Dense::create(this->get_executor(), gko::transpose(this->get_size()));
+    this->conj_transpose(result.get());
+    return result;
 }
 
 
 template <typename ValueType>
-std::unique_ptr<LinOp> Dense<ValueType>::conj_transpose() const
+void Dense<ValueType>::transpose(Dense<ValueType> *output) const
 {
+    GKO_ASSERT_EQUAL_DIMENSIONS(output, gko::transpose(this->get_size()));
     auto exec = this->get_executor();
-    auto trans_cpy = Dense::create(exec, gko::transpose(this->get_size()));
+    exec->run(
+        dense::make_transpose(this, make_temporary_clone(exec, output).get()));
+}
+
 
-    exec->run(dense::make_conj_transpose(this, trans_cpy.get()));
-    return std::move(trans_cpy);
+template <typename ValueType>
+void Dense<ValueType>::conj_transpose(Dense<ValueType> *output) const
+{
+    GKO_ASSERT_EQUAL_DIMENSIONS(output, gko::transpose(this->get_size()));
+    auto exec = this->get_executor();
+    exec->run(dense::make_conj_transpose(
+        this, make_temporary_clone(exec, output).get()));
 }
 
 
 template <typename ValueType>
 std::unique_ptr<LinOp> Dense<ValueType>::permute(
     const Array<int32> *permutation_indices) const
+{
+    auto result = Dense::create(this->get_executor(), this->get_size());
+    this->permute(permutation_indices, result.get());
+    return result;
+}
+
+
+template <typename ValueType>
+std::unique_ptr<LinOp> Dense<ValueType>::permute(
+    const Array<int64> *permutation_indices) const
+{
+    auto result = Dense::create(this->get_executor(), this->get_size());
+    this->permute(permutation_indices, result.get());
+    return result;
+}
+
+
+template <typename ValueType>
+void Dense<ValueType>::permute(const Array<int32> *permutation_indices,
+                               Dense<ValueType> *output) const
 {
     GKO_ASSERT_IS_SQUARE_MATRIX(this);
+    GKO_ASSERT_EQUAL_DIMENSIONS(this, output);
     GKO_ASSERT_EQ(permutation_indices->get_num_elems(), this->get_size()[0]);
     auto exec = this->get_executor();
-    auto permute_cpy = Dense::create(exec, this->get_size());
 
     exec->run(dense::make_symm_permute(
         make_temporary_clone(exec, permutation_indices).get(), this,
-        permute_cpy.get()));
-
-    return std::move(permute_cpy);
+        make_temporary_clone(exec, output).get()));
 }
 
 
 template <typename ValueType>
-std::unique_ptr<LinOp> Dense<ValueType>::permute(
-    const Array<int64> *permutation_indices) const
+void Dense<ValueType>::permute(const Array<int64> *permutation_indices,
```

Comment:
Do you see how we could do this while still keeping (consistent with) the original Permutable interface? Virtual functions can't be templated, so we need to explicitly list them there at least.

Reply:
These `void permute(...)` functions are not virtual, right ? So you should be able to template them ?

Reply:
That would create an inconsistency with the Permutable interface I would like to avoid if possible. I definitely see where you are coming from, I just think having
```
LinOp permute(Array<int32>)
LinOp permute(Array<int64>)
template <typename IndexType>
void permute(Array<IndexType>, Dense)
```
in there might make things more confusing? I wouldn't consider it likely that we ever add another general index type.

Reply:
Yes, but from the user interface it would be the same either way, right. It would reduce duplication from our end, because the functions for the different index types are just doing exactly the same thing ?

Reply:
Maybe what we could do is add internal templated `permute_impl` that are used in these functions? Then the interface stays the same, but we only have a single place where it's implemented?

Reply:
Yes, I think that is a good idea.

Reply:
done